app/controllers/datafiles_controller.rb

```diff
@@ -0,0 +1,172 @@
+class DatafilesController < ApplicationController
+  before_action :set_current_dataset
+
+  def new
+    @datafile = Datafile.new
+
+    if documents?
+      render 'new_doc'
+    end
+  end
+
+  def edit
+    @datafile = current_datafile
+
+    if documents?
+      render 'edit_doc'
+    end
+  end
+
+  def create
+    file_params = params.require(:datafile).permit(:url, :name)
+    @datafile = Datafile.new(file_params)
+    @datafile.dataset = @dataset
+    set_dates(params.require(:datafile).permit(DATE_PARAMS))
+
+    if documents?
+      @datafile.documentation = true
+    end
+
+    if @datafile.save
+      redirect_to files_path(@dataset, new: true) if files?
+      redirect_to documents_path(@dataset) if documents?
+    end
+  end
+
+  def update
+    @datafile = current_datafile
+    file_params = params.require(:datafile).permit(:url, :name)
+    @datafile.update_attributes(file_params)
+    set_dates(params.require(:datafile).permit(DATE_PARAMS))
+
+    if @datafile.save
+      redirect_to files_path(@dataset) if files?
+      redirect_to documents_path(@dataset) if documents?
+    end
+  end
+
+  def destroy
+    @datafile = current_datafile
+    @datafile.destroy
+
+    redirect_to files_path(@dataset) if files?
+    redirect_to documents_path(@dataset) if documents?
+  end
+
+  def index
+    @new = new?
+
+    if documents?
+      documents
+    else
+      files
+    end
+  end
+
+  def files
+    @datafiles = @dataset.datafiles.datalinks
+    render 'files'
+  end
+
+  def documents
+    @datafiles = @dataset.datafiles.documentation
+    render 'documents'
+  end
+
+  private
+  def set_current_dataset
+    @dataset = current_dataset
+  end
+
+  def current_dataset
+    Dataset.find_by(:name => params.require(:id)) || Dataset.find(params.require(:id))
+  end
+
+  def current_datafile
+    Datafile.find(params.require(:file_id))
+  end
+
+  DATE_PARAMS = [
+    :quarter,
+    :year,
+    :start_day,
+    :start_month,
+    :start_year,
+    :end_day,
+    :end_month,
+    :end_year
+  ]
+
+  def set_dates(date_params)
```

Comment:
@govuklaurence If no date is entered when you are adding a link to a datafile there's an error...

<img width="690" alt="screen shot 2017-06-28 at 10 55 38" src="https://user-images.githubusercontent.com/17908089/27631614-f74feee4-5bf0-11e7-9281-21e531a07a68.png">